### ISSUE TITLE
Disconnect VPN when user Kill the app

### ISSUE THREAD

**Issue by SpaceXM** [nizwar/openvpn_flutter#59]
There is any way to perform a VPN disconnection when user kill the app outside my exit button?
Any workaround to avoid VPN stay active also when app is closed?

**Comment by ahmdsdk** [comment]
Having the same issue. Tried engine.disconnect() in dispose() bur it doesn't seem to work. Although it works in reassemble().

**Comment by ahmdsdk** [comment]
@SpaceXM did you find a workaround?

**Comment by ahmdsdk** [comment]
Hi @SpaceXM I just noticed that subscribing to didChangeAppLifecycleState and checking if AppLifecycleState.detached then disconnecting works on Android but not on iOS.

**Comment by 0ttik** [comment]
I don't think this question is in scope of this plugin. You should detect app's closing and call `disconnect` method by yourself. For iOS it's trickier and you should probably use some native capabilities provided by iOS.

**Comment by SpaceXM** [comment]
> Hi @SpaceXM I just noticed that subscribing to didChangeAppLifecycleState and checking if AppLifecycleState.detached then disconnecting works on Android but not on iOS.

i will try this way for android, still searching for IOS

**Comment by Oshaine** [comment]
if you use provider to configure the vpn you will be able to disconnect the app on whatever screen

`  void disconnectVPN(context) async {
    var openVPN = Provider.of<OpenVPNModel>(context, listen: false);
    if (openVPN.connected) {
      print("disconnect");
      openVPN.engine!.disconnect();

      openVPN.connected = false;
    }
  }`

**Comment by Oshaine** [comment]
> > Hi @SpaceXM I just noticed that subscribing to didChangeAppLifecycleState and checking if AppLifecycleState.detached then disconnecting works on Android but not on iOS.
> 
> i will try this way for android, still searching for IOS
> 
> dispose not work just because is called ONLY when user do .pop or close the app using your coded exit button

if you use provider to configure the vpn you will be able to disconnect the app on whatever screen

void disconnectVPN(context) async {
    var openVPN = Provider.of<OpenVPNModel>(context, listen: false);
    if (openVPN.connected) {
      print("disconnect");
      openVPN.engine!.disconnect();

      openVPN.connected = false;
    }
  }

**Comment by syedabdulbasit1** [comment]
@ahmdsdk @SpaceXM did you find anything?, I need urgent help

**Comment by Oshaine** [comment]
This is what i did with the use of provider as well

//globalUtils is a class i created that has my dialog structure

  Future<bool> _onWillPop() async {
    if (_selectedIndex == 1) {
      final dialogResult = await GlobalUtils().openDialog(
          context, 'Are you sure?', 'Do you want to exit the App?', 'No', 'Yes',
          () {
        if (Provider.of<OpenVPNModel>(context, listen: false).connected) {
          OpenVPNConfig().disconnectVPN(context);
        }
        Navigator.of(context).pop(true);
        exit(0);
      });
      // add null check for dialogResult
      if (dialogResult != null) {
        return dialogResult;
      } else {
        return Future.value(false);
      }
    } else {
      final navigation = Provider.of<NavigationModel>(context);
      navigation.selectedIndex = 1;

      return Future.value(true);
    }
  }

**Comment by syedabdulbasit1** [comment]
@Oshaine Thank you, for sharing the solution, but what happens when the user kills the app manually like from the background?
But anyway I just change the app flow when the user kill the app and then restart again then VPN will disconnect.

**Comment by SpaceXM** [comment]
> @Oshaine Thank you, for sharing the solution, but what happens when the user kills the app manually like from the background? But anyway I just change the app flow when the user kill the app and then restart again then VPN will disconnect.

No, is not working in this wat when user kills the app because no fuction can be fired on this event
